### DISCUSSION TITLE
connectOrCreate is great what about connectOrUpsert?

### DISCUSSION THREAD

**Discussion by museadam** [prisma/prisma#25186] [Q&A]
### Question

This is more of a recommendation than a question. I use connectOrCreate a lot, and I run into some situations where I would like to upsert (create or update) a "user" and connect it to a collection. Maybe this can be done this with connectOrCreate but it does not seem like you can. Using prisma/graphQL I've become obsessed with minimizing my database calls and a connectOrUpsert could save me some db calls.

**Comment by ludralph** [comment]
Hi @museadam 👋 

Would a nested upsert work for you as mentioned in this [section](https://www.prisma.io/docs/orm/prisma-client/queries/relation-queries#update-or-create-a-related-record) of the documentation?

You can use nested upsert to update or create a related record:
```js
const result = await prisma.post.update({  
  where: {  
    id: 6,  
  },  
  data: {  
    author: {  
      upsert: {  
        create: {  
          email: 'bob@prisma.io',  
          name: 'Bob the New User',  
        },  
        update: {  
          email: 'bob@prisma.io',  
          name: 'Bob the existing user',  
        },  
      },  
    },  
  },  
  include: {  
    author: true,  
  },  
})
```
In this example, the upsert operation is used to either update an existing user with the email `bob@prisma.io` or create a new user if one does not exist. This approach minimizes the number of database calls by combining the update and create operations into a single query.

**Comment by museadam** [comment]
I cannot nest a upsert inside a connectOrCreate.

this is one example with your upsert suggestion and I get this error .
`Object literal may only specify known properties, and 'upsert' does not exist in type 'UserCreateOrConnectWithoutApplicationsInput`


` res = await prisma.application.create({

			data: {

				jobPost: {

					connect: {
						id: jobId
					}
				},
				resume: {
					...resumObj
				},
				applicant: {
					connectOrCreate: {
						where: {
							id: userId ?? ''
						},
						upsert: {
							create: {
								email,
								firstName,
								lastName,
								name,
								eeoAnswers,
								resumes: {
									create: {
										...resumeData
									}
								}
							},
							update: {
								eeoAnswers
							}
						}
					}
				}
			},

			include: {
				applicant: {
					include: {
						resumes: true
					}
				}
			}
		});`

**Reply by ludralph** [reply]
Hi @museadam, 

Can you please share your schema? Nesting an `upsert` inside a `connectOrCreate` is not supported. The `upsert` query allows you to update an existing record or create a new one if it doesn't exist, which can help minimize your database calls. Here's  how you can use the `upsert` query to update or create a User record and connect it to related records:

```js
const user = await prisma.user.upsert({
  where: { id: 1 },  
  update: { 
    email: 'alice@prisma.io',
    posts: {  
      connectOrCreate: [  
        {  
          where: { id: 32 },  
          create: { title: 'This is my first post' },  
        },  
        {  
          where: { id: 19 },  
          create: { title: 'This is my second post' },  
        },  
      ],  
    },
  },
  create: { 
    email: 'alice@prisma.io',
    posts: {  
      connectOrCreate: [  
        {  
          where: { id: 32 },  
          create: { title: 'This is my first post' },  
        },  
        {  
          where: { id: 19 },  
          create: { title: 'This is my second post' },  
        },  
      ],  
    },
  },
})
```
The `upsert` query is used to either update an existing `User` record or create a new one if it doesn't exist. The `connectOrCreate` nested query is used to connect the user to existing `Post` records or create new Post records if they don't exist.

**Comment by museadam** [comment]
Sir you told me to do a nested upsert and I was showing you that it can not be done. I had already known it doesn't work, that is why in my original post I recommended that a connectOrUpsert be considered for future releases.

**Reply by willbjones** [reply]
Did you find a work around for this, I found this limitation almost immediately when adopting prisma.. I guess you upsert all of your child records prior to then connecting them to the parent.

**Reply by ludralph** [reply]
Hi @museadam 

Can you please open a [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) to see this supported in the future?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;

**Comment by Telokis** [comment]
In case someone finds this in the future, the following feature request might deserve your upvote: https://github.com/prisma/prisma/issues/7787
